### DISCUSSION TITLE
Customising python packages included in apps

### DISCUSSION THREAD

**Discussion by mwhapples** [beeware/briefcase#1594] [Q&A]
I am creating an app using briefcase. For the UI I am using PySide6. I am only using some of the PySide modules, however briefcase is including all of PySide6 in my app. This leads to a very large app installer and large installation size (nearly 500MB on Windows). Is there any way to get briefcase to exclude some of the packages? Other tools like nuitka, PyInstaller and cx_Freeze allow this or do a certain amount automatically. For the same app building with nuitka my bundle is about 76MB.

**Comment by rmartin16** [comment]
You can limit the packages included in the final bundle with `requires` in `pyproject.toml`.

Perhaps more what you're looking for, though, you can remove select paths from the bundle via [`cleanup_paths`](https://briefcase.readthedocs.io/en/stable/reference/configuration.html#cleanup-paths) in the `briefcase.toml`. For instance, AppImage builds [remove](https://github.com/beeware/briefcase-linux-appimage-template/blob/main/%7B%7B%20cookiecutter.format%20%7D%7D/briefcase.toml) several components of the Python install; the same could be done with the PySide6 install.

**Reply by mwhapples** [reply]
Thanks for the reply. The requires option is not going to help as it deals with packages from pypi, what I need is to exclude parts of those packages. If my code never imports PySide6.QtLocation, I need not have the QtLocation.pyd and related files. However I need other modules from PySide6 so it must be in the requires variable. Yes cleanup_paths may be able to do what I need, although it may be more verbose as I need to work out the unused dependencies of what I do not import as well.
This may be the answer, but for the moment I will leave not marked as answered in case there is another option which may be better suited to what I need. If something could automatically attempt the dependency detection and filtering by inspecting the imports of my code that would be great.

**Reply by freakboy3742** [reply]
Fundmentally, this is something you need to take up with Qt or the PySide maintainers. Briefcase is limited in what it can do here - it can install a package that is provided on PyPI; and it can delete files once they're installed. There isn't really a third option here - either the actual packages being installed need to be smaller, or you delete files from a large package.

It sounds like what you're looking for is a Python equivalent of "Tree shaking" - the idea of downloading everything, then automatically deleting the parts that "aren't used". This has been proposed in the past - see #215 for the feature request - but it's suspiciously difficult to do in practice, because of how dynamic Python is as a language. It's not that hard to parse all the Python code to extract all the modules that are mentioned in `import` statements; but that will miss any call to `importlib`, or any import executed inside extension code - something that Qt will be especially prone to, as very little of it's code is actually written in Python. 

At the end of the day, tree shaking is essentially the automated form of `cleanup_paths` - but it turns out to be more reliable in the Python context.
